core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -1146,7 +1146,7 @@ public RelBuilder functionScan(SqlOperator operator,
    * and optimized in a similar way to the {@link #and} method.
    * If the result is TRUE no filter is created. */
   public RelBuilder filter(RexNode... predicates) {
-    return filter(ImmutableList.copyOf(predicates));
+    return filter(ImmutableList.copyOf(predicates), Collections.emptyList());
```

Comment:
Instead of doing a copy, `Arrays.asList()` would work as well (copy would be done later by filter constructor)

Reply:
`ImmutableList.copyOf` is better than `Arrays.asList` or `new ArrayList()`. The next method will see that it is an `ImmutableList` and know that it does not need to make a further copy.

Furthermore, it is slightly better to use `ImmutableList.of()` rather than `Collections.emptyList()` because then we are using Guava ImmutableList everywhere.

Reply:
True if the argument is used as-is all the way long, but if `filter(Iterable<? extends RexNode>, Iterable<CorrelationId>)` does need to transform/alter the argument, it might an unnecessary copy.

Reply:
Actually, if it's going to end up as an `ImmutableSet` eventually, let's use `ImmutableSet.copyOf`. (Another reason to like `ImmutableSet` is that it preserves iteration order. If we were to use `HashSet` as the intermediate form it would make tests non-deterministic.)

Reply:
The copy might be unnecessary, but it's not easy for the called method to know whether to make a defensive copy, because it has public and private callers.

So the safe/sane thing to do is to get to immutable collections as early in the process as possible.

Reply:
I think Laurent is mentioning predicates argument which is converted into a single RexNode condition in the builder, and the one that is passed along and end up as an ImmutableSet is correlation variables. So, I think we could use ImmutableSet for correlation variables and, for predicates, we can pass a list. Updated in the new PR.